creator-node/src/utils.js

```diff
@@ -71,6 +72,72 @@ async function getIPFSPeerId (ipfs, config) {
   return ipfsIDObj
 }
 
+async function rehydrateIpfsFromFsIfNecessary (req, multihash, storagePath, filename = null) {
+  let ipfs = req.app.get('ipfsAPI')
+  let ipfsPath = multihash
+  if (filename != null) {
+    // Indicates we are retrieving a directory multihash
+    ipfsPath = `${multihash}/${filename}`
+  }
+
+  // Confirm availability by catting single byte locally
+  let timeoutPromise = new Promise((resolve, reject) => {
+    let wait = setTimeout(() => {
+      clearTimeout(wait)
+      reject(new Error('TIMEOUT'))
+    }, 1000)
+  })
+
+  let ipfsSingleByteCat = new Promise(async (resolve, reject) => {
+    // Cat single byte
+    const start = Date.now()
+    await ipfs.cat(ipfsPath, { length: 1 })
+    console.log(`rehydrateIpfsFromFsIfNecessary - Retrieved ${ipfsPath} in ${Date.now() - start}ms`)
+    // retrieved = true
+    resolve('SUCCESS')
+  })
+
+  let rehydrateNecessary = false
+  try {
+    await Promise.race([
+      timeoutPromise,
+      ipfsSingleByteCat])
+  } catch (e) {
+    rehydrateNecessary = true
+    console.log(`rehydrateIpfsFromFsIfNecessary - error condition met ${ipfsPath}, ${e}`)
+  }
+  if (!rehydrateNecessary) return
+  // Timed out, must re-add from FS
+  if (!filename) {
+    console.log(`rehydrateIpfsFromFsIfNecessary - Re-adding file - ${multihash}, stg path: ${storagePath}`)
+    let addResp = await ipfs.addFromFs(storagePath, { pin: false })
+    console.log(`rehydrateIpfsFromFsIfNecessary - Re-added file - ${multihash}, stg path: ${storagePath},  ${JSON.stringify(addResp)}`)
+  } else {
+    console.log(`rehydrateIpfsFromFsIfNecessary - Re-adding dir ${multihash}, stg path: ${storagePath}, filename: ${filename}, ipfsPath: ${ipfsPath}`)
+    let findOriginalFileQuery = await models.File.findAll({
+      where: {
+        storagePath: { [models.Sequelize.Op.like]: `%${multihash}%` },
+        type: 'image'
+      }
+    })
+    // Add entire directory to recreate original operation
+    // Required to ensure same dirCID as data store
+    let ipfsAddArray = []
+    for (var entry of findOriginalFileQuery) {
+      let sourceFilePath = entry.storagePath
+      let bufferedFile = fs.readFileSync(sourceFilePath)
+      let originalSource = entry.sourceFile
+      ipfsAddArray.push({
+        path: originalSource,
+        content: bufferedFile
+      })
+    }
+    let addResp = await ipfs.add(ipfsAddArray, { pin: false })
```

Comment:
do we need to also ipfs.add the dir CID here, similar to the `/image_upload` route? would it make sense to just re-use that code here?
```
await ipfs.add([
          { path: path.join(req.file.originalname, '2000x.jpg'), content: imageBuffer2000x },
          { path: path.join(req.file.originalname, '640x.jpg'), content: imageBuffer640x },
          { path: path.join(req.file.originalname, 'original.jpg'), content: imageBufferOriginal }
        ], { pin: false })
```

Reply:
the object in ipfsAddArray is recreating exactly that no? we just don't have all the buffers at once here and query them / iterate

Reply:
yes you right resolved